### PR TITLE
Fix Insufficient Balance to Pay Ticket Fee

### PR DESCRIPTION
**Pull request by oshorefueled**
This PR resolves issue #176 

It adds some of the methods used in dcrwallet vsp package to reserve some inputs for paying for fees after puchasing a ticket. It adds a vsp package which consists of some helper functions and client code for making requests to the vsp server

### CODE REVIEW

File: types.go [hunk 1]
```diff
@@ -379,16 +391,16 @@ type GetFeeAddressResponse struct {
...
 }
 
 type PayFeeResponse struct {
-	Timestamp int64           `json:"timestamp"`
-	Request   json.RawMessage `json:"request"`
+	Timestamp int64  `json:"timestamp"`
+	Request   []byte `json:"request"`
 }
 
 type TicketStatusRequest struct {
```
Comment (C-ollins) [inline types.go:406]: Remove unused structs. 
- TicketStatusRequest
- TicketStatusResponse
- SetVoteChoicesRequest
- SetVoteChoicesResponse
- GetFeeAddressRequest
- GetFeeAddressResponse

File: vsp.go [hunk 2]
```diff
@@ -0,0 +1,526 @@
...
+	pubKey          []byte
+	w               *Wallet
+	mwRef           *MultiWallet
+	purchaseAccount uint32
+	changeAccount   uint32
+	chainParams     *chaincfg.Params
+
+	*vsp.Client
+	ctx context.Context
+	c   *w.ConfirmationNotificationsClient
```
Comment (C-ollins) [inline vsp.go:41]: This variable is not used, remove this & other unused variables in this struct.

File: vsp.go [hunk 3]
```diff
@@ -0,0 +1,526 @@
...
+		return -1, err
+	}
+	return vspInfo.FeePercentage, nil
+}
+
+// PurchaseStakeTickets purchases the number of tickets passed as an argument and pays the fees for the tickets
+func (v *VSP) PurchaseTickets(ticketCount, expiryBlocks int32, passphrase []byte) error {
+	defer v.w.LockWallet()
+	passphraseCopy := make([]byte, len(passphrase))
+	_ = copy(passphraseCopy, passphrase)
```
Comment (C-ollins) [inline vsp.go:104]: Why is this done?

File: vsp.go [hunk 4]
```diff
@@ -0,0 +1,526 @@
...
+	err := v.Client.Get(ctx, apiVSPInfo, &vspInfo)
+	if err != nil {
+		return -1, err
+	}
+	return vspInfo.FeePercentage, nil
+}
+
+// PurchaseStakeTickets purchases the number of tickets passed as an argument and pays the fees for the tickets
+func (v *VSP) PurchaseTickets(ticketCount, expiryBlocks int32, passphrase []byte) error {
+	defer v.w.LockWallet()
```
Comment (C-ollins) [inline vsp.go:102]: Move this to after successfully unlocking the wallet.

File: vsp.go [hunk 5]
```diff
@@ -0,0 +1,526 @@
...
+	}
+
+	networkBackend, err := v.w.internal.NetworkBackend()
+	if err != nil {
+		return err
+	}
+
+	_, err = v.w.internal.PurchaseTickets(v.ctx, networkBackend, request)
+	if err != nil {
+		log.Errorf("PURCHASE TICKET %v", err.Error())
```
Comment (C-ollins) [inline vsp.go:127]: This log is not needed since the error is returned, to avoid spamming the logs.

File: vsp.go [hunk 6]
```diff
@@ -0,0 +1,526 @@
...
+
+	err = v.Client.Post(ctx, "/api/v3/payfee", feeInfo.CommitmentAddress,
+		&payfeeResponse, json.RawMessage(requestBody))
+	if err != nil {
+		return err
+	}
+
+	// verify initial request matches server
+	if !bytes.Equal(requestBody, payfeeResponse.Request) {
+		log.Warnf("[PAYFee] server response has differing request: %#v != %#v",
```
Comment (C-ollins) [inline vsp.go:380]: Remove spam logs...

**Review by C-ollins — CHANGES_REQUESTED** [review]

File: vsp.go (hunk 6 above)
Comment (oshorefueled) [reply vsp.go:380]: I'm not sure if this should be considered spam log. There are a number of places where requests are made to the vsp server and they return this error. Logging it with the differences would make it easier to debug and fix. The error message returned "server response contains differing request" is a user friendly message that could easily be used by the caller of the method.

**Review by C-ollins — CHANGES_REQUESTED** [review]

**Review by C-ollins — APPROVED** [review]

**Pull request merged**